File: _week-1-solution.py
import geopandas as gpd

# File paths
gpkg = "data/swisstlm3d_2024-03_2056_5728/SWISSTLM3D_2024_LV95_LN02.gpkg"
gpkg2 = "data/swissboundaries3d_2024-01_2056_5728/swissBOUNDARIES3D_1_5_LV95_LN02.gpkg"

# Load layers
bb = gpd.read_file(gpkg, layer="tlm_bb_bodenbedeckung")
cantons = gpd.read_file(gpkg2, layer="tlm_kantonsgebiet")

# Filter forest areas
wald = bb[bb["objektart"] == "Wald"]

# Intersect forest areas with canton boundaries
wald_cant = gpd.overlay(wald, cantons, how="intersection")

# Calculate forest area
wald_cant["area"] = wald_cant.geometry.area
wald_area = wald_cant.groupby("name", as_index=False)["area"].sum().rename(columns={"area": "wald_area"})

# Calculate total canton area
cantons["area"] = cantons.geometry.area
canton_area = cantons.groupby("name", as_index=False)["area"].sum().rename(columns={"area": "canton_area"})

# Merge and compute forest percentage
result = wald_area.merge(canton_area, on="name", how="outer")
result["wald_perc"] = (result["wald_area"] / result["canton_area"]) * 100

# Sort by forest percentage in descending order
result = result.sort_values(by="wald_perc", ascending=False)

# Display result
print(result)

File: _week-1-solution.R
library(sf)
library(dplyr)


gpkg <- "data/swisstlm3d_2024-03_2056_5728/SWISSTLM3D_2024_LV95_LN02.gpkg"
gpkg2 <- "data/swissboundaries3d_2024-01_2056_5728/swissBOUNDARIES3D_1_5_LV95_LN02.gpkg"


bb <- st_read(gpkg, "tlm_bb_bodenbedeckung")
cantons <- st_read(gpkg2, "tlm_kantonsgebiet")

wald <- bb |> filter(objektart == "Wald")


wald_cant <- st_intersection(wald, cantons)

wald_cant$area <- as.numeric(st_area(wald_cant))

wald_area <- wald_cant |> 
  st_drop_geometry() |> 
  group_by(name) |> 
  summarise(
    wald_area = sum(area)
  )

cantons$area <- as.numeric(st_area(cantons))

canton_area <- cantons |> 
  st_drop_geometry() |> 
  group_by(name) |> 
  summarise(
    canton_area = sum(area)
  )

full_join(wald_area, canton_area) |> 
  mutate(wald_perc = wald_area/canton_area*100) |> 
  arrange(desc(wald_perc))

